server/src/main/java/org/apache/lucene/index/BaseStarTreeBuilder.java

```diff
@@ -0,0 +1,626 @@
+ */
+package org.apache.lucene.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.DocValuesConsumer;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.index.compositeindex.datacube.DateDimension;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.compositeindex.datacube.Metric;
+import org.opensearch.index.compositeindex.datacube.MetricStat;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeField;
+import org.opensearch.index.compositeindex.datacube.startree.StarTreeFieldConfiguration;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.MetricStatFieldPair;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregator;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.ValueAggregatorFactory;
+import org.opensearch.index.compositeindex.datacube.startree.aggregators.numerictype.StarTreeNumericType;
+import org.opensearch.index.compositeindex.datacube.startree.builder.SingleTreeBuilder;
+import org.opensearch.index.compositeindex.datacube.startree.builder.StarTreeDocValuesIteratorAdapter;
+import org.opensearch.index.compositeindex.datacube.startree.data.StarTreeDocument;
+import org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeBuilderUtils;
+import org.opensearch.index.fielddata.IndexNumericFieldData;
+import org.opensearch.index.mapper.Mapper;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.NumberFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for star-tree builder
+ */
+public abstract class BaseStarTreeBuilder implements SingleTreeBuilder {
+
+    private static final Logger logger = LogManager.getLogger(BaseStarTreeBuilder.class);
+
+    public static final int STAR_IN_DOC_VALUES_INDEX = -1;
+
+    protected final String[] dimensionsSplitOrder;
+    protected final Set<Integer> skipStarNodeCreationForDimensions;
+    protected final String[] metrics;
+
+    protected final int numMetrics;
+    protected final int numDimensions;
+    protected int numStarTreeDocs;
+    protected int totalSegmentDocs;
+    protected int numStarTreeNodes;
+    protected final int maxLeafDocuments;
+
+    protected final StarTreeBuilderUtils.TreeNode rootNode = getNewNode();
+
+    protected DocIdSetIterator[] dimensionReaders;
+    protected DocIdSetIterator[] metricReaders;
+
+    protected ValueAggregator[] valueAggregators;
+    protected IndexNumericFieldData.NumericType[] numericTypes;
+    protected DocValuesConsumer docValuesConsumer;
+    protected DocValuesProducer docValuesProducer;
```

Comment:
I'll provide a `Map<String, DocValuesProducer>` and `Map<String, FieldInfo>` to this class to fetch original segment doc values.

Reply:
ack! would be making this change